src/data/sequence.rs

```diff
@@ -0,0 +1,727 @@
+pub type Values = Vec<Value>;
+
+/// A representation of data at some version.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Default, Debug)]
+pub struct DataEntry {
+    /// Data version
+    pub version: u64,
+    /// Data value
+    pub value: Vec<u8>,
+}
+
+impl DataEntry {
+    /// Returns a new instance of a data entry.
+    pub fn new(version: u64, value: Vec<u8>) -> Self {
+        Self { version, value }
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash)]
+pub struct SequenceBase<P, S> {
+    address: Address,
+    data: Values,
+    access_list: Vec<P>,
+    // This is the history of owners, with each entry representing an owner.  Each single owner
+    // could represent an individual user, or a group of users, depending on the `PublicKey` type.
+    owners: Vec<Owner>,
+    _flavour: S,
+}
+
+/// Common methods for all `Sequence` flavours.
+impl<C, S> SequenceBase<C, S>
+where
+    C: AccessListTrait,
+    S: Copy,
+{
+    /// Returns true if the provided access type is allowed for the specific user (identified y their public key).
+    pub fn is_allowed(&self, user: PublicKey, access: AccessType) -> bool {
+        if let Some(owner) = self.owner_at(CURRENT_VERSION) {
+            if owner.public_key == user {
+                return true;
+            }
+        }
+        match self.access_list_at(CURRENT_VERSION) {
+            Some(list) => list.is_allowed(&user, access),
+            None => false,
+        }
+    }
+
+    /// Return the address of this Sequence.
+    pub fn address(&self) -> &Address {
+        &self.address
+    }
+
+    /// Return the name of this Sequence.
+    pub fn name(&self) -> &XorName {
+        self.address.name()
+    }
+
+    /// Return the type tag of this Sequence.
+    pub fn tag(&self) -> u64 {
+        self.address.tag()
+    }
+
+    /// Returns true if the user is the current owner.
+    pub fn is_owner(&self, user: PublicKey) -> bool {
+        match self.owner_at(CURRENT_VERSION) {
+            Some(owner) => user == owner.public_key,
+            _ => false,
+        }
+    }
+
+    /// Return the expected data version.
+    pub fn expected_data_version(&self) -> u64 {
+        self.data.len() as u64
+    }
+
+    /// Return the expected owners version.
+    pub fn expected_owners_version(&self) -> u64 {
+        self.owners.len() as u64
+    }
+
+    /// Return the expected access list version.
+    pub fn expected_access_list_version(&self) -> u64 {
+        self.access_list.len() as u64
+    }
+
+    pub fn versions(&self) -> ExpectedVersions {
+        ExpectedVersions::new(
+            self.expected_data_version(),
+            self.expected_owners_version(),
+            self.expected_access_list_version(),
+        )
+    }
+
+    /// Returns the data shell - that is - everything except the Values themselves.
+    pub fn shell(&self, expected_data_version: impl Into<Version>) -> Result<Self> {
```

Comment:
I'm not sure what's the role of `expected_data_version` here?
Is this to fetch the data shell at a particular version of the data?

Reply:
> Is this to fetch the data shell at a particular version of the data?

Yes. This logic has not changed from before. Only the name of the parameter.

Reply:
If you are talking of the shell method parameter, then that's a simple rename.

The other methods in the highlighted code, deal with that version that we by convention call expected version.

Reply:
Yea we do. It's just that "expected version" doesn't seem to fit in a get operation.
We'll need to define some stuff clearly. 
Eg. Do versions start from 0 or 1?
When we insert a new value for the first time, is it version 0 or version 1?
I think it's worth getting this clarified and do it right this time.

Reply:
> "expected version" doesn't seem to fit in a get operation

It depends. As a parameter, where you are asking for some actual data, then yes, it should just be Version. Definitely.
If you are asking for the expected version (i.e. what version is expected?) Then.. you do a GET and the returned result are the "expected versions".

> We'll need to define some stuff clearly.
> Eg. Do versions start from 0 or 1?

What you request is part of the previous PR I did ([as seen here](https://github.com/maidsafe/safe-nd/pull/126)), when we introduced current and expected version convention. All that is included here. Versions always start at 0.

Reply:
> the returned result are the "expected versions".
Yes it makes sense for a result. But as a paramter is what I'm not sure of.

This function is to return the shell of the data. So `data.shell(0)` is get shell at version 0 `data.shell(1)` is get shell at version 1. Which makes sense. 

But right now it's implemented as `data.shell(1)` which interprets as get the shell of the data when the next expected version is 1. Which might be confusing right?

Reply:
Yeah absolutely, that must be some remnant, line 538 has it correct, the parameter is named `version` there.

It's just the name of the parameter that is wrong, the logic is as it should (you'd pass in 0 and get at version 0).